File: lib/NativeGL/gl_window.rb
$: << File.expand_path(File.dirname(__FILE__))
require 'ffi'

module CWrapper
	extend FFI::Library
	ffi_lib 'libwin.so'
	
	attach_function :loadLibrary, [], :int
	attach_function :createGLWindow, [:int, :int, :int, :int, :int, :int, 
	                                  :string, :uint32], :pointer
	attach_function :createGLContext, [:pointer], :int
	attach_function :setGLAttributes, [:int, :int, :int, :int], :void
										
	attach_function :swapBuffer, [:pointer], :void
										
	attach_variable :FULLSCREEN, :WINDOW_FULLSCREEN, :int
	attach_variable :SHOW, :WINDOW_SHOWN, :int
	attach_variable :HIDE, :WINDOW_HIDDEN, :int
	attach_variable :BORDERLESS, :WINDOW_BORDERLESS, :int
	attach_variable :RESIZABLE, :WINDOW_RESIZABLE, :int
	attach_variable :MINIMIZE, :WINDOW_MINIMIZED, :int
	attach_variable :MAXIMIZE, :WINDOW_MAXIMIZED, :int
	attach_variable :INPUT_FOCUS, :WINDOW_INPUT_FOCUS, :int
	attach_variable :MOUSE_FOCUS, :WINDOW_MOUSE_FOCUS, :int
	attach_variable :FULLSCREEN_DESKTOP, :WINDOW_FULLSCREEN_DESKTOP, :int
	attach_variable :HIGHDPI, :WINDOW_ALLOW_HIGHDPI, :int
end

module RubyWrapper
	def self.initGLEW()
		if (CWrapper.initGLEW == 0) then
			raise 'Was not able to initialize GLEW.'
		end
	end
	
	def self.createGLWindow(options = {})
		options = { major: 3, minor: 3, xPosition: 100, yPosition: 100, 
		            width: 480, height: 480, title: 'LibraryName', flags: 0
		          }.merge(options)
		
		CWrapper.createGLWindow(options[:major], options[:minor], 
		                        options[:xPosition], options[:yPosition], 
		                        options[:width], options[:height], 
		                        options[:title], options[:flags])
	end
	
	def self.createGLContext(glWindow)
		CWrapper.createGLContext(glWindow)
	end
end

File: lib/NativeGL/src/GLWindow.c
#include "GLWindow.h"

const SDL_WindowFlags WINDOW_FULLSCREEN = SDL_WINDOW_FULLSCREEN,
                      WINDOW_OPENGL = SDL_WINDOW_OPENGL,
                      WINDOW_SHOWN = SDL_WINDOW_SHOWN, 
                      WINDOW_HIDDEN = SDL_WINDOW_HIDDEN,
                      WINDOW_BORDERLESS = SDL_WINDOW_BORDERLESS, 
                      WINDOW_RESIZABLE = SDL_WINDOW_RESIZABLE, 
                      WINDOW_MINIMIZED = SDL_WINDOW_MINIMIZED,
                      WINDOW_MAXIMIZED = SDL_WINDOW_MAXIMIZED, 
                      WINDOW_INPUT_GRABBED = SDL_WINDOW_INPUT_GRABBED, 
                      WINDOW_INPUT_FOCUS = SDL_WINDOW_INPUT_FOCUS,
                      WINDOW_MOUSE_FOCUS = SDL_WINDOW_MOUSE_FOCUS, 
                      WINDOW_FULLSCREEN_DESKTOP = SDL_WINDOW_FULLSCREEN_DESKTOP, 
                      WINDOW_FOREIGN = SDL_WINDOW_FOREIGN,
                      WINDOW_ALLOW_HIGHDPI = SDL_WINDOW_ALLOW_HIGHDPI;

void setGLAttributes(int majorVersion, int minorVersion, int depthSize, 
                     int doubleBuffer)
{
	SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, majorVersion);
	SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, minorVersion);

	SDL_GL_SetAttribute(SDL_GL_DEPTH_SIZE, depthSize);
	SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, doubleBuffer);
}

SDL_Window* createGLWindow(int majorVersion, int minorVersion, int xPosition, 
                           int yPosition, int width, int height, 
                           const char* title, Uint32 windowFlags)
{
	setGLAttributes(majorVersion, minorVersion, 24, 1);

	SDL_Window* mainWindow = SDL_CreateWindow(
		title,
		xPosition,
		yPosition,
		width,
		height,
		windowFlags | SDL_WINDOW_OPENGL
	);

	return mainWindow;
}

SDL_GLContext createGLContext(SDL_Window* window)
{
	SDL_GLContext mainContext = SDL_GL_CreateContext(window);


	SDL_GL_MakeCurrent(window, mainContext);
	SDL_GL_SetSwapInterval(1);

	return mainContext;
}

Uint32 getCombinedFlags(int makeFullscreen, int makeFullscreenDesktop, 
                        int makeHidden, int makeBorderless, int makeResizable, 
                        int makeMinimized, int makeMaximized, int makeFocus, 
                        int makeHDPI)
{
	Uint32 flags = 0;
	
	if (makeFullscreen)
		flags |= SDL_WINDOW_FULLSCREEN;
		
	if (makeFullscreenDesktop)
		flags |= SDL_WINDOW_FULLSCREEN_DESKTOP;
		
	if (makeHidden)
		flags |= SDL_WINDOW_HIDDEN;
		
	if (makeBorderless)
		flags |= SDL_WINDOW_BORDERLESS;
		
	if (makeResizable)
		flags |= SDL_WINDOW_RESIZABLE;
		
	if (makeMinimized)
		flags |= SDL_WINDOW_MINIMIZED;
		
	if (makeMaximized)
		flags |= SDL_WINDOW_MAXIMIZED;
		
	if (makeFocus)
		flags |= SDL_WINDOW_INPUT_GRABBED;
		
	if (makeHDPI)
		flags |= SDL_WINDOW_ALLOW_HIGHDPI;
		
	return flags;
}